src/libsyntax/parse/parser/path.rs

```diff
@@ -470,4 +497,50 @@ impl<'a> Parser<'a> {
 
         Ok((args, constraints))
     }
+
+    fn recover_bare_const_expr(&mut self) -> PResult<'a, P<Expr>> {
+        let snapshot = self.clone();
+        debug!("recover_bare_const_expr {:?}", self.token);
+        match self.parse_expr_res(Restrictions::CONST_EXPR_RECOVERY, None) {
+            Ok(expr) => {
+                debug!("recover_bare_const_expr expr {:?} {:?}", expr, expr.node);
+                if let token::Comma | token::Gt = self.token.kind {
+                    // We parsed the whole const argument successfully without braces.
+                    debug!("recover_bare_const_expr ok");
+                    if !expr.node.is_valid_const_on_its_own() {
+                        // But it wasn't a literal, so we emit a custom error and
+                        // suggest the appropriate code. `foo::<-1>()` is valid but gets parsed
+                        // here, so we need to gate the error only for invalid cases.
+                        self.span_fatal(
+                            expr.span,
+                            "complex const arguments must be surrounded by braces",
+                        ).multipart_suggestion(
+                            "surround this const argument in braces",
+                            vec![
+                                (expr.span.shrink_to_lo(), "{ ".to_string()),
+                                (expr.span.shrink_to_hi(), " }".to_string()),
+                            ],
+                            Applicability::MachineApplicable,
+                        ).emit();
+                    }
+                    Ok(expr)
```

Comment:
The method is a bit of a misnomer now because `Ok(expr)` here is not necessarily an error branch?